src/gan/gen_pyt/model/parser.py

```diff
@@ -889,10 +893,51 @@ def parse(self, src_sent, context=None, beam_size=5, debug=False):
             else:
                 break
 
+        if return_states:
+            saved_states = [x for _, x in 
+                            sorted(zip(completed_hypotheses, saved_states), 
+                                   key=lambda pair: -pair[0].score)]
+
         completed_hypotheses.sort(key=lambda hyp: -hyp.score)
 
+        if return_states:
+            return completed_hypotheses, saved_states
+        else:
+            return completed_hypotheses
+
         return completed_hypotheses
 
+    def sample(self, src_sent, hyp=None, states=None):
+        """Perform beam search to infer the target AST given a source utterance
+           and optionally an incomplete Hypothesis.
+
+        Args:
+            src_sent: list of source utterance tokens
+            hyp: incomplete hypothesis
+            states: history of decoder states for generating Hypothesis
+            beam_size: beam size
+
+        Returns:
+            A completed Hypothesis
+        """
+
+        if hyp is not None and hyp.completed:
```

Comment:
I'm not following what's going on here. Why is the double statement necessary? And, below, what's the purpose of the `while True`... what happens that's different the second time the `while` loop is executed that didn't happen the first time (or third, or fourth, etc.)?